python/ml4ir/applications/ranking/features/feature_fns/string.py

```diff
@@ -63,4 +65,96 @@ def call(self, inputs, training=None):
+    NUMERIC_REGEX = r"^.*[0-9]+.*$"
+    PUNCTUATION_REGEX = "^.*[" + "".join([re.escape(c) for c in list(string.punctuation)]) + "]+.*$"
+
+
+    VOCABULARY = [
+        ALPHA_QUERY_TYPE,
+        NUMERIC_QUERY_TYPE,
+        PUNCTUATION_QUERY_TYPE,
+        ALPHA_QUERY_TYPE + NUMERIC_QUERY_TYPE,
+        ALPHA_QUERY_TYPE + PUNCTUATION_QUERY_TYPE,
+        NUMERIC_QUERY_TYPE + PUNCTUATION_QUERY_TYPE,
+        ALPHA_QUERY_TYPE + NUMERIC_QUERY_TYPE + PUNCTUATION_QUERY_TYPE
+    ]
+
+    def __init__(self, feature_info: dict, file_io: FileIO, **kwargs):
+        """
+        Initialize layer to define a query type vectorizer
+
+        Parameters
+        ----------
+        feature_info : dict
+            Dictionary representing the configuration parameters for the specific feature from the FeatureConfig
+        file_io : FileIO object
+            FileIO handler object for reading and writing
+
+        Notes
+        -----
+        Args under feature_layer_info:
+            remove_quotes : string
+                Whether to remove quotes from the string tensors
+                Defaults to true
+            output_mode : str
+                the type of vector representation to compute
+                currently supports either embedding or one_hot
+            embedding_size : int
+                dimension size of categorical embedding
+        """
+        super().__init__(feature_info=feature_info, file_io=file_io, **kwargs)
+
+        self.remove_quotes = self.feature_layer_args.get(self.REMOVE_QUOTES, True)
+
+        feature_info["feature_layer_info"]["args"]["vocabulary"] = self.VOCABULARY
+        self.categorical_vector_op = CategoricalVector(feature_info, file_io, **kwargs)
+
+    def call(self, inputs, training=None):
+        """
+        Defines the forward pass for the layer on the inputs tensor
+
+        Parameters
+        ----------
+        inputs: tensor
+            Input tensor on which the feature transforms are applied
+        training: boolean
+            Boolean flag indicating if the layer is being used in training mode or not
+
+        Returns
+        -------
+        tf.Tensor
+            Resulting tensor after the forward pass through the feature transform layer
+        """
+        # Replace quotes and white spaces
+        if self.remove_quotes:
+            inputs = tf.strings.regex_replace(inputs, """["']""", "")
+
+        # Assign query type
+        query_type = tf.strings.join([
+            tf.where(tf.strings.regex_full_match(inputs, self.ALPHA_REGEX), self.ALPHA_QUERY_TYPE, ""),
+            tf.where(tf.strings.regex_full_match(inputs, self.NUMERIC_REGEX), self.NUMERIC_QUERY_TYPE, ""),
+            tf.where(tf.strings.regex_full_match(inputs, self.PUNCTUATION_REGEX), self.PUNCTUATION_QUERY_TYPE, ""),
+        ])
+
+        # Vectorize the query type to either dense embedding or sparse one-hot
+        query_type_vector = self.categorical_vector_op(query_type, training=training)
```

Comment:
Converting the query type string feature into either one-hot or an embedding feature to be used in the model.